File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/execution/ReplaceMatchesEvaluator.java
package org.opencds.cqf.cql.engine.elm.execution;

import org.opencds.cqf.cql.engine.execution.Context;

public class ReplaceMatchesEvaluator extends org.cqframework.cql.elm.execution.ReplaceMatches {

    public static Object replaceMatches(String argument, String pattern, String substitution) {
        if (argument == null || pattern == null || substitution == null) {
            return null;
        }

        return argument.replaceAll(pattern, substitution);
    }

    @Override
    protected Object internalEvaluate(Context context) {
        String argument = (String) getOperand().get(0).evaluate(context);
        String pattern = (String) getOperand().get(1).evaluate(context);
        String substitution = (String) getOperand().get(2).evaluate(context);

        return replaceMatches(argument, pattern, substitution);
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/ReplaceMatchesEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import kotlin.jvm.JvmStatic
import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument
import org.opencds.cqf.cql.engine.runtime.CqlType
import org.opencds.cqf.cql.engine.runtime.String
import org.opencds.cqf.cql.engine.runtime.toCqlString

object ReplaceMatchesEvaluator {
    @JvmStatic
    fun replaceMatches(argument: CqlType?, pattern: CqlType?, substitution: CqlType?): String? {
        if (argument == null || pattern == null || substitution == null) {
            return null
        }

        if (argument is String && pattern is String && substitution is String) {

            return argument.replace(pattern.value.toRegex(), substitution.value).toCqlString()
        }

        throw InvalidOperatorArgument(
            "ReplaceMatches(String, String, String)",
            "ReplaceMatches(${argument.typeAsString}, ${pattern.typeAsString}, ${substitution.typeAsString})",
        )
    }
}
